File: Trees/kth_smallest_val.py
# https://leetcode.com/problems/kth-smallest-element-in-a-bst

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        self.k = k
        self.result = 0
        self.helper(root)
        return self.result

    def helper(self, root):
        if root is None:
            return
        self.helper(root.left)
        self.k -= 1
        if self.k == 0:
            self.result = root.val
        self.helper(root.right)

File: Trees/kth_smallest_val.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}

def kth_smallest(root, k)
  # Inorder traversal?
  @elements = []
  inorder(root)
  @elements[k - 1]
end

def inorder(root)
  return if root == nil
  
  inorder(root.left)
  @elements << root.val
  inorder(root.right)
end



# Alternative

def kth_smallest(root, k)
  # Inorder traversal?
  @k = k
  @element = nil
  inorder(root, 0)
  @element
end

def inorder(root, k)
  return k + 1 if root == nil
  
  position = inorder(root.left, k)
  @element = root.val if position == @k
  inorder(root.right, position) # 2, 1
end